### PR TITLE
add retain flag to pub

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

Add `retain` option while publishing the messages.
Now one can publish as `tedge mqtt pub --retain topic message`

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [x ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [x] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: crates/core/tedge/src/cli/mqtt/cli.rs [hunk 1]
```diff
@@ -21,6 +21,9 @@ pub enum TEdgeMqttCli {
         /// QoS level (0, 1, 2)
         #[structopt(short, long, parse(try_from_str = parse_qos), default_value = "0")]
         qos: QoS,
+        /// Retain flag
+        #[structopt(long = "retain")]
```
Comment (albinsuresh) [inline crates/core/tedge/src/cli/mqtt/cli.rs:25]: A short option `-r` would also be nice.

**Review by albinsuresh — COMMENTED** [review]
Can one of the tedge mqtt pysys tests be enhanced or a new one added to test this feature?

File: crates/core/tedge/src/cli/mqtt/cli.rs (hunk 1 above)
Comment (PradeepKiruvale) [reply crates/core/tedge/src/cli/mqtt/cli.rs]: Yes, added.

File: crates/core/tedge/src/cli/mqtt/cli.rs [hunk 2]
```diff
@@ -21,6 +21,9 @@ pub enum TEdgeMqttCli {
         /// QoS level (0, 1, 2)
         #[structopt(short, long, parse(try_from_str = parse_qos), default_value = "0")]
         qos: QoS,
+        /// Retain flag
+        #[structopt(short, long = "retain")]
```
Comment (didier-wenzek) [inline crates/core/tedge/src/cli/mqtt/cli.rs:25]: We are missing a default value here (should be false).

File: crates/core/tedge/src/cli/mqtt/publish.rs [hunk 3]
```diff
@@ -32,7 +33,11 @@ impl Command for MqttPublishCommand {
 fn publish(cmd: &MqttPublishCommand) -> Result<(), MqttError> {
     let mut options = MqttOptions::new(cmd.client_id.as_str(), &cmd.host, cmd.port);
     options.set_clean_session(true);
-    let retain_flag = false;
+    let mut retain_flag = false;
+
+    if cmd.retain {
+        retain_flag = true;
+    }
```
Comment (didier-wenzek) [inline crates/core/tedge/src/cli/mqtt/publish.rs:40]:
```suggestion
retain_flag = cmd.retain
```

You can even directly use `cmd.retain` to publish the message.

```
client.publish(&cmd.topic, cmd.qos, cmd.retain, payload)?;
```

File: docs/src/howto-guides/005_pub_sub.md [hunk 4]
```diff
@@ -17,6 +17,13 @@ tedge mqtt pub 'tedge/measurements' '{ "temperature": 21.3 }'
 ```shell
 tedge mqtt pub 'tedge/measurements' '{ "temperature": 21.3 }' --qos 2
 ```
+`tedge mqtt pub` supports publishing a MQTT message with retain flag:
```
Comment (didier-wenzek) [inline docs/src/howto-guides/005_pub_sub.md:20]: I would add a blank line.

File: docs/src/howto-guides/005_pub_sub.md [hunk 5]
```diff
@@ -17,6 +17,13 @@ tedge mqtt pub 'tedge/measurements' '{ "temperature": 21.3 }'
...
+
+```shell
+tedge mqtt pub --retain 'tedge/measurements' '{ "temperature": 21.3 }'
+```
```
Comment (didier-wenzek) [inline docs/src/howto-guides/005_pub_sub.md:24]: It's better to have an example that makes sense

```suggestion
tedge mqtt pub --retain --qos 1 tedge/alarms/critical/high_temperature '{"message": "Temperature is critical"}'
```

File: docs/src/howto-guides/005_pub_sub.md [hunk 6]
```diff
@@ -17,6 +17,13 @@ tedge mqtt pub 'tedge/measurements' '{ "temperature": 21.3 }'
...
+Note: By default the mqtt message will be published with retain flag set to false.
```
Comment (didier-wenzek) [inline docs/src/howto-guides/005_pub_sub.md:25]: Add a blank line between the code sample and the note.

File: crates/core/tedge/src/cli/mqtt/cli.rs (hunk 2 above)
Comment (PradeepKiruvale) [reply crates/core/tedge/src/cli/mqtt/cli.rs:25]: Tried that, but got an error "message: default_value is meaningless for bool".

File: docs/src/howto-guides/005_pub_sub.md (hunk 6 above)
Comment (PradeepKiruvale) [reply docs/src/howto-guides/005_pub_sub.md:27]: fixed

File: docs/src/howto-guides/005_pub_sub.md (hunk 5 above)
Comment (PradeepKiruvale) [reply docs/src/howto-guides/005_pub_sub.md]: updated

File: docs/src/howto-guides/005_pub_sub.md (hunk 4 above)
Comment (PradeepKiruvale) [reply docs/src/howto-guides/005_pub_sub.md:21]: fixed

File: crates/core/tedge/src/cli/mqtt/publish.rs (hunk 3 above)
Comment (PradeepKiruvale) [reply crates/core/tedge/src/cli/mqtt/publish.rs:40]: done.

**Review by didier-wenzek — APPROVED** [review]
Approved

**Pull request merged**